### PR TITLE
[RLlib] Fix bugs in IMPALA/APPO + LSTM (new stack) and activate StatelessCartPole learning tests on new API stack.

### PR DESCRIPTION
**Pull request by sven1977**
Fix bugs in IMPALA/APPO + LSTM (new stack) and activate StatelessCartPole learning tests on new API stack.

* `AddOneTsToEpisodesAndTruncate` connector needs to disseminate each episode chunk to NOT add an extra timestep to the middle of an episode (which happens to have two chunks in the training data).
* Fix APPO learner bugs when training with RNNs (would crash due to a malformed batch).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/appo/torch/appo_torch_learner.py [hunk 1]
```diff
@@ -54,8 +54,9 @@ def compute_loss_for_module(
         )
         behaviour_actions_logp = batch[Columns.ACTION_LOGP]
         target_actions_logp = target_policy_dist.logp(batch[Columns.ACTIONS])
+
         rollout_frag_or_episode_len = config.get_rollout_fragment_length()
-        recurrent_seq_len = None
```
Comment (simonsays1980) [inline rllib/algorithms/appo/torch/appo_torch_learner.py:58]: Ah, that cannot work, of yourse.

File: rllib/algorithms/impala/torch/vtrace_torch_v2.py [hunk 2]
```diff
@@ -31,14 +31,16 @@ def make_time_major(
...
+        trajectory_len is not None or recurrent_seq_len is not None
     ), "Either trajectory_len or recurrent_seq_len must be set."
 
     # Figure out the sizes of the final B and T axes.
-    if recurrent_seq_len:
-        B = recurrent_seq_len.shape[0]
-        T = tensor.shape[0] // B
+    if recurrent_seq_len is not None:
+        assert len(tensor.shape) == 2
+        # Swap B and T axes.
```
Comment (simonsays1980) [inline rllib/algorithms/impala/torch/vtrace_torch_v2.py:41]: As the customer asked today: is this conversion needed? Isn't it done by a connector? Our base LSTM encoder does not do it as far as I remember.

File: rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py [hunk 3]
```diff
@@ -101,10 +101,24 @@ def __call__(
         # batch: - - - - - - - T B0- - - - - R Bx- - - - R Bx
         # mask : t t t t t t t t f t t t t t t f t t t t t f
 
-        shared_data["_sa_episodes_lengths"] = {}
-        for sa_episode in self.single_agent_episode_iterator(
-            episodes, agents_that_stepped_only=False
+        for i, sa_episode in enumerate(
+            self.single_agent_episode_iterator(episodes, agents_that_stepped_only=False)
         ):
+            # TODO (sven): This is a little bit of a hack: By expanding the Episode's
```
Comment (simonsays1980) [inline rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py:107]: Tricky :/

File: rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py [hunk 4]
```diff
@@ -101,10 +101,24 @@ def __call__(
...
+            #  ID, we make sure that each episode chunk in `episodes` is treated as a
+            #  separate episode in the `self.add_n_batch_items` below. Some algos (e.g.
+            #  APPO) may have >1 episode chunks from the same episode (same ID) in the
+            #  training data, thus leading to a malformatted batch in case of
+            #  RNN-triggered zero-padding of the train batch.
+            #  For example, if e1 (id=a len=4) and e2 (id=a len=5) are two chunks of the
+            #  same episode in `episodes`, the resulting batch would have an additional
+            #  timestep in the middle of the episode's "row":
+            #  {  "obs": {
```
Comment (simonsays1980) [inline rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py:116]: Couldn't we just concat the episode and then cut them into sequence len pieces?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Stateful modules in APPO/IMPALA is tricky. I would like to see also a note in the docs and a comment in the APPORLModule how using an LSTM should be handled by users. I am sure many will have similar questions as our customer. 

In addition: how would this look like when a user uses an attention model in her module? Does she also have to override the VTrace calculation or any connector?

File: rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py (hunk 4 above)
Comment (sven1977) [reply rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py:116]: Yeah, I was thinking about that, too. But I think this would cost too much extra time? It would certainly be of advantage b/c it would eliminate the "bad" samples, where we have very short episode chunks (a new episode started close to the end (e.g. 2 timesteps away) of the EnvRunner's rollout fragment length) and then zero-pad that trajectory all the way to the end, where instead we could concatenate that short piece with the longer successor piece to give it "more meaning".

We could try adding this logic to this same connector (for now) and run some experiments with CartPole (generally short episodes) and Pong (generally quite long episodes) to see, whether there is a performance delta ...

File: rllib/algorithms/impala/torch/vtrace_torch_v2.py (hunk 2 above)
Comment (sven1977) [reply rllib/algorithms/impala/torch/vtrace_torch_v2.py:41]: It's a conversion specifically needed for v-trace, which sits inside the APPO/IMPALA loss. So I think it's a good design to make the loss itself do this flip (instead of the connector). Also, we only need to flip a few columns, not all of them.
